### PR TITLE
Better support HLS media discontinuity and fully support media initialization

### PR DESCRIPTION
**Pull request by shirt-dev**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/pukkandan/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

Better support HLS media discontinuity and media initialization.
Old behavior can be restored with `--hls-ignore-discontinuity`

### CODE REVIEW

**Comment by shirt-dev** [comment]
Note: test_parse_m3u8_formats tests will need to be updated in test_InfoExtractor.py

File: youtube_dlc/extractor/common.py [hunk 1]
```diff
@@ -1868,6 +1870,8 @@ def _parse_m3u8_formats(self, m3u8_doc, m3u8_url, ext=None,
             if re.match(r'^https?://', u)
             else compat_urlparse.urljoin(m3u8_url, u))
 
+        ignore_discontinuity = self._downloader.params.get('hls_ignore_discontinuity')
```
Comment (pukkandan) [inline youtube_dlc/extractor/common.py:1873]:
```suggestion
        ignore_discontinuity = self._downloader.params.get('hls_ignore_discontinuity', True)
```

Keep the old behavior by default. We don't want any scripts/URLs that used to work correctly to break

File: youtube_dlc/options.py [hunk 2]
```diff
@@ -654,6 +654,11 @@ def _dict_from_multiple_values_options_callback(
         help=(
             'Use the mpegts container for HLS videos, allowing to play the '
             'video while downloading (some players may not be able to play it)'))
+    downloader.add_option(
+        '--hls-ignore-discontinuity',
+        dest='hls_ignore_discontinuity', action='store_true', default=None,
+        help='Ignore media discontinuity on HLS playlists (do not split ad breaks)'
+    )
```
Comment (pukkandan) [inline youtube_dlc/options.py:661]:
```suggestion
    downloader.add_option(
        '--hls-split-discontinuity',
        dest='hls_ignore_discontinuity', action='store_false', default=True,
        help='Split discontinuity on HLS playlists to different formats'
    )
    downloader.add_option(
        '--no-hls-split-discontinuity',
        dest='hls_ignore_discontinuity', action='store_true',
        help='Concatenate media discontinuity on HLS playlists (default)'
    )
```

Feel free to change the actual switch name and description.

**Review by pukkandan — COMMENTED** [review]
I don't know enough about the actual parsing to review it thoroughly. I trust that you know what you are doing

* Don't remove the existing tests (only add new ones)
* Keep default behavior as-is
* Give switches for both enabling and disabling this behavior so that user can manually override any option given in a config file
* I think the switches should be in `Extractor Options`, not downloader options

**Pull request merged**